iseq.c

```diff
@@ -248,230 +245,155 @@ rb_iseq_each_value(const rb_iseq_t *iseq, iseq_value_itr_t * func, void *data)
 
-                    if (vm_cc_markable(cc)) {
-                        if (!vm_cc_invalidated_p(cc)) {
-                            rb_gc_mark_movable((VALUE)cc);
-                        }
-                        else {
-                            cds[i].cc = rb_vm_empty_cc();
-                        }
+                    if (reference_updating || !vm_cc_invalidated_p(cds[i].cc)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].cc);
+                    }
+                    else {
+                        cds[i].cc = rb_vm_empty_cc();
                     }
                 }
             }
         }
 
         if (body->param.flags.has_kw && ISEQ_COMPILE_DATA(iseq) == NULL) {
             const struct rb_iseq_param_keyword *const keyword = body->param.keyword;
-            int i, j;
-
-            i = keyword->required_num;
 
-            for (j = 0; i < keyword->num; i++, j++) {
-                VALUE obj = keyword->default_values[j];
-                if (!SPECIAL_CONST_P(obj)) {
-                    rb_gc_mark_movable(obj);
+            for (int j = 0, i = keyword->required_num; i < keyword->num; i++, j++) {
+                if (!SPECIAL_CONST_P(keyword->default_values[j])) {
+                    rb_gc_mark_and_move(&keyword->default_values[j]);
                 }
             }
         }
 
         if (body->catch_table) {
-            const struct iseq_catch_table *table = body->catch_table;
-            unsigned int i;
-            for (i = 0; i < table->size; i++) {
-                const struct iseq_catch_table_entry *entry;
+            struct iseq_catch_table *table = body->catch_table;
+
+            for (unsigned int i = 0; i < table->size; i++) {
+                struct iseq_catch_table_entry *entry;
                 entry = UNALIGNED_MEMBER_PTR(table, entries[i]);
                 if (entry->iseq) {
-                    rb_gc_mark_movable((VALUE)entry->iseq);
+                    rb_gc_mark_and_move_ptr(&entry->iseq);
                 }
             }
         }
 
+        if (reference_updating) {
+#if USE_MJIT
+            mjit_update_references(iseq);
+#endif
+#if USE_YJIT
+            rb_yjit_iseq_update_references(body->yjit_payload);
+#endif
+        }
+        else {
 #if USE_MJIT
-        mjit_mark_cc_entries(body);
+            mjit_mark_cc_entries(body);
 #endif
 #if USE_YJIT
-        rb_yjit_iseq_mark(body->yjit_payload);
+            rb_yjit_iseq_mark(body->yjit_payload);
 #endif
+        }
     }
 
-    if (FL_TEST_RAW((VALUE)iseq, ISEQ_NOT_LOADED_YET)) {
-        rb_gc_mark(iseq->aux.loader.obj);
-    }
-    else if (FL_TEST_RAW((VALUE)iseq, ISEQ_USE_COMPILE_DATA)) {
-        const struct iseq_compile_data *const compile_data = ISEQ_COMPILE_DATA(iseq);
+    // TODO: make these not pinned
+    if (!reference_updating) {
+        if (FL_TEST_RAW((VALUE)iseq, ISEQ_NOT_LOADED_YET)) {
+            rb_gc_mark(iseq->aux.loader.obj);
```

Comment:
Alternatively, add a `bool pin` parameter to `rb_gc_mark_and_move` to indicate whether it needs to pin a field.  If true, the GC can pin it when called during the mark phase, and not forward it during the forwarding (compaction) phase.  For evacuating GC algorithms, it may be interested in recording all children visited by `rb_gc_mark_and_move(..., pin=true)` before GC starts.

Reply:
I don't plan on doing that, since the next step after this PR will be to look into removing all of the `rb_gc_mark` in this function and make everything movable.